remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,804 @@
+    index++;
+    offset += len;
+  }
+}
+
+enum class VisitResult {
+  Progress,
+  NoProgress,
+  Error
+};
+
+enum class OpType {
+  Plus,
+  Minus,
+};
+
+// Get the unsigned offset of two int64_t numbers with bounds checking
+static bool GetUnsignedOffset(int64_t v1, int64_t v2, OpType op, int64_t max, uint64_t *result) {
+  auto signed_result = v1;
+  switch (op) {
+    case OpType::Plus:
+      signed_result += v2;
+      break;
+    case OpType::Minus:
+      signed_result -= v2;
+      break;
+    default:
+      break;
+  }
+  if (signed_result >= 0 && signed_result < max) {
+    *result = static_cast<uint64_t>(signed_result);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+ForwardAliasVisitor::ForwardAliasVisitor(
+    const std::vector<StateSlot> &state_slots_, llvm::DataLayout *dl_, llvm::Value *sp_)
+    : state_slots(state_slots_),
+      state_offset({{sp_, 0}}),
+      state_access_offset(),
+      exclude(),
+      curr_wl(),
+      state_ptr(sp_),
+      dl(dl_) {}
+
+void ForwardAliasVisitor::AddInstruction(llvm::Instruction *inst) {
+  curr_wl.insert(inst);
+}
+
+// Iterate through the current worklist, updating the state_offset and state_access_offset
+// according to the instructions in the list. Any instruction that is not
+// currently interpretable (some of its pointers are not yet in the state_offset)
+// is withheld to the next analysis round in the next worklist.
+// Analysis repeats until the current worklist is empty.
+bool ForwardAliasVisitor::Analyze(void) {
+  bool progress = true;
+  std::unordered_set<llvm::Instruction *> next_wl;
+  // if any visit makes progress, continue the loop
+  while (!curr_wl.empty() && progress) {
+    progress = false;
+    for (auto inst : curr_wl) {
+      switch (visit(inst)) {
+        case VisitResult::Progress:
+          progress = true;
+          break;
+        case VisitResult::NoProgress:
+          next_wl.insert(inst);
+          break;
+        case VisitResult::Error:
+          return false;
+      }
+    }
+    // update curr_wl to the next set of insts
+    curr_wl.swap(next_wl);
+    // reset upcoming insts
+    next_wl.clear();
+  }
+  //TODO(tim): check for possible errors?
+  return true;
+}
+
+VisitResult ForwardAliasVisitor::visitInstruction(llvm::Instruction &I) {
+  return VisitResult::Progress;
+}
+
+VisitResult ForwardAliasVisitor::visitAllocaInst(llvm::AllocaInst &I) {
+  exclude.insert(&I);
+  return VisitResult::Progress;
+}
+
+// Visit a load instruction and update the alias map.
+VisitResult ForwardAliasVisitor::visitLoadInst(llvm::LoadInst &I) {
+  // clear any aametadata
+  llvm::AAMDNodes aamd(nullptr, nullptr, nullptr);
+  I.setAAMetadata(aamd);
+  // get the initial ptr
+  auto val = I.getPointerOperand();
+  // special case: loading the state ptr
+  if (I.getType() == state_ptr->getType()) {
+    state_offset.emplace(&I, 0);
+    return VisitResult::Progress;
+  } else if (exclude.count(val)) {
+    exclude.insert(&I);
```

Comment:
Actually, let's not do this. My bad, it could pessimize our `visitBinaryOp_` analysis!

Reply:
Duplicate? Or do you never want to update the `exclude` set from `visitLoadInst(&I)`?

Reply:
Oh hrmm, actually now that I think of it again, we *do* want to add to exclude set. At first I thought this would screw up the `visitBinaryOp_`, but then I realized there's no way to convert a loaded value, whose value isn't known, into a constantint ;-)